### ISSUE TITLE
Debugger breakpoints: wrong bank

### ISSUE THREAD

**Issue by andrew-davie** [stella-emu/stella#512]
It seems that when you have a breakpoint, only the PC address is used/recorded for the breakpoint. If I am stepping through code and I click in the breakpoint column to add a breakpoint, and then continue - the breakpoint can be (is!) triggered by access to the PC address from a different bank. This is not good behaviour, because I don't want to break on code in a different bank - I want to break on code in the bank I set the breakpoint in!

In other words, breakpoints should keep track of the bank they belong to. Yes, this isn't possible to know if you set a breakpoint via label ("break label"), but in that case you would want to break when PC = label in ANY bank. But when you do it manually, you're stepping through code, you most definitely do NOT want it to break whenever PC=addr.  You want it to break ONLY when PC==addr AND bank=bank that was switched in when breakpoint was set.

In other words, I think breakpoints also need a bank field. Normally this should mean "any bank" but in the case where set via step-through and set-breakpoint-here type usage, then the bank should be constrained to the one you are currently in.

**Comment by sa666666** [comment]
Also related to https://github.com/stella-emu/stella/issues/217.

**Comment by thrust26** [comment]
We could use the already existing conditional breakpoints (e.g. via config or in case of a multi bank ROM).

Currently all breakpoints only check for the PC. This is done via a 64k bitset for speed reasons. But maybe this is not that relevant anymore?

**Comment by andrew-davie** [comment]
Keep using the 64k bitset for initial check, but once a breakpoint is discovered, then check if it is the correct bank and if not, ignore...?

**Comment by sa666666** [comment]
I've though about this in the past, but never had time to revisit it.  It also needs work in the UI, since clicking an address in the disassembly right now creates a _naive_ breakpoint (ie, one that only looks at address, not bank).  There needs to be a way to graphically set the kind you want (address AND bank).  And perhaps have them be different colours/shapes too, to indicate which is active.

I agree with @thrust26 that the easiest way with our current infrastructure is to just use a CBP.  So the former looks like this:
``` break $f050```
while the CBP would be like this:
```breakif {_pc==$f050 && _bank==x}``` (where 'x' is the bank we want)

So the infrastructure is there, just with a little more typing.  Presenting a choice to the user in UI mode would then allow the code to 'write' that breakpoint for us.  In the end, I think that's the best approach.

**Comment by thrust26** [comment]
> Keep using the 64k bitset for initial check, but once a breakpoint is discovered, then check if it is the correct bank and if not, ignore...?

Maybe. A problem are multiple breakpoints at the same address. We would have to add a counter for each address. Else deleting one breakpoint would affect the other ones too.

Maybe something for later, if the new solution turns out to be too slow.

**Comment by sa666666** [comment]
I still think a CBP is what we need.  The bitset is for a simple use-case: unconditional breakpoint.  As soon as we add an extra _condition_ (ie, a corresponding bank), then it becomes a _conditional_ breakpoint, hence CBP.  The infrastructure is already there; there's no need to complicate the bitset approach to do something it wasn't designed to do.  Just use a CBP and be done with it.  Also, they're not _that_ slow.

**Comment by sa666666** [comment]
In any event, the UI work still needs to be done no matter what approach we take.  It should be easy to just point and click, and add the different types of BP without having to type anything.  But until that happens, the typing approach works right now.

**Comment by thrust26** [comment]
Agreed. I will have a look.

**Comment by sa666666** [comment]
Well, I didn't mean to imply that it had to be _you_, or even that we need to do this before 6.1.  I just wanted to point out that this is more of a UI issue than anything else.  The bankend can already do it; we just need to extend the functionality to the frontend.

**Comment by thrust26** [comment]
If we completely remove the unconditional breakpoints, we have no UI issue anymore.

**Comment by sa666666** [comment]
There are two issues with going that route:
- the unconditional BP are faster; but maybe that doesn't matter so much anymore?
- more importantly, what if someone want to set a BP and _not_ consider banks?  Again, maybe that isn't a use case we need to worry about??

**Comment by sa666666** [comment]
I defer to the experience of people actually using Stella for real 2600 development; you and @andrew-davie.

**Comment by andrew-davie** [comment]
I have been following the conversation. I agree with the proposed solution regarding using existing infrastructure (i.e., CBP) to implement the breakpoints with bank. That makes perfect sense to me. I don’t really see how anyone would want a breakpoint WITHOUT a bank, so I’d probably drop that capability.

**Comment by thrust26** [comment]
And one could still define a conditional breakpoint manually, without providing the bank as argument.

**Comment by andrew-davie** [comment]
So, when I’m in the UI and stepping through… and I double-click on the column for breakpoints, it can put in an automatic “banked” breakpoint for me (as a CBP), because it knows the bank it’s in, and of course the PC. No need for a menu or anything.  As to using the command-line, if I type “break label” then perhaps it could add a CPB breakpoint for EVERY bank at that label address. Then I get the original behaviour. If that’s not what I wanted I can disable each breakpoint that I didn’t want via UI as I hit it.

**Comment by sa666666** [comment]
> As to using the command-line, if I type “break label” then perhaps it could add a CPB breakpoint for EVERY bank at that label address. Then I get the original behaviour. If that’s not what I wanted I can disable each breakpoint that I didn’t want via UI as I hit it.

This is probably the most elegant solution.  If we make _all_ BP conditional, then the whole problem goes away (UI issues included).  Then the only potential issue is that CBP are slower.  But we can test that right now, before any work is started.  I don't think it will be an issue on any modern computer, and even if it slows down slightly, that's to be expected when enabling debugging features.

**Comment by thrust26** [comment]
@andrew-davie  Yup. 

And you can type `break label` or `break label bank`. At least that's the plan.

Though removing break points added without bank by clicking in the breakpoint column needs some thought.: 
- We can either add N breakpoints initially and then you remove one. 
- Or we add one breakpoint without a bank. Then, if the first one is removed via breakpoints column, we can either remove that global breakpoint. Or we remove that one and replace it with N-1 breakpoints in the other banks. 

What's better?

**Comment by sa666666** [comment]
This address mirroring in the 2600 really complicates things, doesn't it? :smiling_imp:

**Comment by thrust26** [comment]
@sa666666 You won't be surprised it I tell you that I already started? :smile:

One more thought: "break label" could alternatively add the current bank. Then the problem described above is gone.

**Comment by sa666666** [comment]
> @sa666666 You won't be surprised if I tell you that I already started? I love challenges! 

Sigh.  I used to feel that way too, and I miss it.  I'm just so overloaded with stuff now that I'm getting a little burnt out.  Hopefully I can get back to the 'challenging' stuff soon too.  Multi-bank disassembly is one that I really want to work on ...

**Comment by thrust26** [comment]
When I am back at work next Tuesday, I will more like you.

Multi-bank disassembly will become very challenging, DiStella was never meant for that and the code shows. I happily leave that task for up. 😈 

BTW: Is there an issue for this?

**Comment by sa666666** [comment]
Yes, https://github.com/stella-emu/stella/issues/284.  In fact you created it!

**Comment by thrust26** [comment]
I just checked the maximum frame rates with multiple conditional breakpoints. The slowdown is quite noticeable on my notebook (i7-7820HQ):
- 0 breakpoints >600 frames/s
- 5 breakpoints: ~585 frames/s
- 10 breakpoints: ~530 frames/s
- 15 breakpoints: ~465 frames/s
- 20 breakpoints: ~425 frames/s

So for slower machines we should consider some optimizations.

**Comment by sa666666** [comment]
I don't think that's actually that bad.  How many times would one have 20 BP at a time?  And even if the machine is 10x slower and gets 60fps max, now maybe it only gets 40fps max (assuming this scales linearly).  Of course I'm all for efficiency, but at the same time, if you're doing heavy debugging then you have to accept more CPU usage and potential slowdown.

**Comment by thrust26** [comment]
Yea, the single thread performance of my Z8350 based mini notebook is only about 5x slower than the i7. So that should be sufficient.

And I did some minor optimizations (e.g. no bank check for non-bankswitched ROMs).

**Comment by sa666666** [comment]
You can probably also extend that optimization to ROMs that support only one bank but _still_ have bankswitching.  Some of the more complicated schemes only report one bank, since it's much too complicated otherwise.  In those cases the bank check is also redundant.

**Comment by thrust26** [comment]
But couldn't their banks share the same addresses?

**Comment by sa666666** [comment]
What I mean for the schemes is that it's redundant to ask whether the current bank matches the one that was set when the BP was created.  There is only _one_ bank, so it's useless to do the check.

**Comment by thrust26** [comment]
Oh, that's what I meant to say. I check the cart's bank count. 

@andrew-davie Since you are coding quite a lot, maybe you can test the debugger a bit?

**Comment by andrew-davie** [comment]
I’ll be very happy to test. I did download and built Stella on MacOS some weeks ago, but without Xcode. It worked, much to my surprise, but the fonts/visuals weren’t quite as good as the official release. I guess I should install Xcode and do it properly!

One issue I have had with debugging such large programs split over many banks is the labels that the disassembly chooses to show. It would be great to be able to click on an erroneous label (in a particular bank) and go something like “not this one” or perhaps have a droplets of labels with the same value as the PC at the right-click location and let me choose the correct one. And then assign a bank to that label so it is always displayed in the correct bank and no others.

Same with variables. If there are multiple (I use overlays, so there are always multiples everywhere) then perhaps show that with a slightly different colour or indicator that there are other options, and a droplets to let me choose. It would just make debugging so much easier if the labels/variables were actually showing the correct thing.

**Comment by thrust26** [comment]
It helps with the labels if you define different ORGs for each bank.

Labels are more a problem. I doubt even overlays etc. would be very helpful, there are just too many initial wrong "guesses". There should be some better logic, but that would involve using the DASM files. And that would be a major rewrite of the disassembly, which is currently based in DiStella code with many of its flaws still left.

**Comment by thrust26** [comment]
@andrew-davie Not using XCode might explain why you dislike the "Light" theme that much. :smile:

**Comment by andrew-davie** [comment]
No, I wasn’t clear. I tried compiling my own version and didn’t use it. I was just confirming I CAN compile it without Xcode.
I dislike the light theme in the official release :)

**Comment by DirtyHairy** [comment]
@andrew-davie There shouldn't be any visual differences between "proper" XCode and CLI builds. The only difference is that the CLI builds use the "generic" Unix entrypoint and thus

 * Miss the "About" dialog
 * Use the same paths as Linux for storing stuff
 * Save settings as an .ini file (or to a SQlite DB) instead of MacOS style application properties

Also, I am not sure whether the serial port works, as MacOS port has its own serial port driver. However, I wouldn't be surprised if even that works with the CLI build.

In a nutshell, the CLI build should be functionally identical to the XCode buils, I am using it all the time. It is just different from the "official" MacOS built, that's why we put the warning message into the `configure` script.

**Comment by andrew-davie** [comment]
Thank you for the follow-up on this.
I am seeing differences between the head build and the previous official release, and that was why I did not like what I was seeing.
You are correct I have now done an Xcode build and it looks the same as the CLI. But there are still issues. I will file a bug report.
The issues are related to visible banding on the Stella screen which I believe are caused by scaling artefacts.
Not present on the previous release, and very noticeable in the current build.

**Comment by andrew-davie** [comment]
This is a comment on the latest build, 6.1_beta2, with regard to breakpoints...
I have done some preliminary testing. The first thing that struck me...

now in the command window you have to type...

"break label bank#"

This is not what I thought we agreed upon. In the above example, we have broken backward-compatibility with how people are used to setting breakpoints. In particular this becomes a major pain/inconvenience when you are using a non-bankswitched game.

What I thought we agreed was when you type...

"break label"

... then it will add CPBs for ALL banks at the label's PC. If you (optionally) decide to type in a bank number then it adds only one CPB for the designated bank.

**Comment by andrew-davie** [comment]
The new <bank#> option on the console breakpoint setting defaults to hexadecimal.
This is really offputting. I type "break reset 15" and it sets the breakpoint to 21.
It TELLS me it sets to 21. There's a disconnect between what I type and what I see, because it assumes I am typing hex, yet tells me what I typed in decimal. I believe the default should be decimal, and if you want to specify hex, put $15.
If it has to be hex, then can we have consistency in the output (i.e., in hex, too), and also in the help specify that the bank needs to be in hexadecimal?
but i'd prefer decimal, as default!!
<img width="320" alt="Screen Shot 2019-08-21 at 12 08 49 pm" src="https://user-images.githubusercontent.com/16424075/63397041-992c9c00-c40c-11e9-9ff0-7ea4dab9061b.png">

**Comment by andrew-davie** [comment]
Breakpoints are definitely.... broken.
I identify the bank of a label "Reset" (which is $F, in this case - the fixed bank). I then run and trigger the debugger, and add the breakpoint "break Reset F". Then I reset and run, and a breakpoint triggers in bank 11 at $F016. I list the breakpoints - they are as expected, a single one at 'Reset" PC address $FBA1 (correct, according to the symbol table). So, there's a bogus break happening in the wrong bank at the wrong PC address. I have a screen grab/video of this in actino, but it seems I can't attach MP4s.
I've found this build is just not working well with breakpoints at all. Also I'm getting a lot of outright crashes, though I am not sure if that's related to the breakpoint stuff, or my build.

**Comment by andrew-davie** [comment]
Furthermore, I believe in all debugging output, the radix should be very clear. If you are showing values as hex, then identify them as hex ($val). OR make sure you never never ever mix radix such as in the above display... it should say "set breakpoint at fbb3 in bank f" not i "in bank 15".
But, in a perfect world, it would actually say... "set breakpoint at $fbb3 in bank $f"

**Comment by thrust26** [comment]
Confirmed. Aaargh!

**Comment by thrust26** [comment]
> ... then it will add CPBs for ALL banks at the label's PC. If you (optionally) decide to type in a bank number then it adds only one CPB for the designated bank.

This will be fixed with the following commit.

**Comment by thrust26** [comment]
> but i'd prefer decimal, as default!!

You have to use `#` as prefix for decimals, all numbers are hex by default in the debugger and I don't want to mix that case by case.

**Comment by thrust26** [comment]
> Breakpoints are definitely.... broken.
> I identify the bank of a label "Reset" (which is $F, in this case - the fixed bank). I then run and trigger the debugger, and add the breakpoint "break Reset F". Then I reset and run, and a breakpoint triggers in bank 11 at $F016. I list the breakpoints - they are as expected, a single one at 'Reset" PC address $FBA1 (correct, according to the symbol table). So, there's a bogus break happening in the wrong bank at the wrong PC address. I have a screen grab/video of this in actino, but it seems I can't attach MP4s.
> I've found this build is just not working well with breakpoints at all. Also I'm getting a lot of outright crashes, though I am not sure if that's related to the breakpoint stuff, or my build.

Yup, there was a problem with banks larger than 9. I tried to enforce using decimals there, but that didn't work well. So back to hexadecimals for now. 

BTW: I am now logging break expression bugs to the console, please let me know if you spot one. They caused the crashed to were experiencing.

**Comment by thrust26** [comment]
@andrew-davie Just committed a fix. I tested with MegaBoy, because it has >= 10 banks, so I hope I got all bugs. Please let me know if everything is OK now or if you finder further problems.

Fixed: 
- bank number parsing for banks >= 10
- output now with preceding '$'
- `break` works without providing bank number

**Comment by andrew-davie** [comment]
No luck. Just did a fetch and build, and tried the following...

Enter debugger
Type "break Reset".  I would expect to see it add 16 CPBs.
However it just adds one, in a seemingly random bank.
I then type "break Reset F" because that's where I want it.
Then "reset" then "run".  NO breakpoints are triggered :(
I list the breakpoints, just the two mentioned...
I confirmed the "Reset" label is $FBA1 and bank $F
The red 'breakpoint dot' *DOES* appear at the correct spot in the disassembly, it just doesn't seem to trigger.

<img width="400" alt="Screen Shot 2019-08-21 at 8 07 31 pm" src="https://user-images.githubusercontent.com/16424075/63423355-77083d80-c44f-11e9-8c4a-816b08fd9c70.png">

**Comment by andrew-davie** [comment]
To be clear, 'break Reset' should be equivalent to typing...

break Reset 0
break Reset 1
break Reset 2
break Reset 3
break Reset 4
.... all the way up to...
break Reset F

in other words, it should add 16 CBPs.
Only in the case of a non-banked rom should break Reset be equivalent to 'break Reset 0'.
In other words, again, I don't think just adding one CBP is correct - you want to catch the label as a breakpoint in ALL banks, if none are specified.

**Comment by thrust26** [comment]
`break` and `break addr` currently automatically add the current bank **only**. IMO that makes more sense then adding all banks (which also would make me worry about too many breakpoints and the resulting potential slowdowns with current code). Or do you want to make a difference between e.g. `break f000` and `break mylabel`? Where the latter adds N breakpoints and the former only one. That seems overly complicated.

> The red 'breakpoint dot' DOES appear at the correct spot in the disassembly, it just doesn't seem to trigger.

Can you double check this, please? I cannot reproduce that bug.

**Comment by sa666666** [comment]
Concerning the mixing of bases (hex, decimal, etc).  If we want to work on this, then please move it to a separate issue.  This thread is starting to get complicated and going in all different directions.  The discussion here should be for the `bank` functionality only.  Hex vs. decimal input is a separate issue.

**Comment by sa666666** [comment]
And we need to clearly list of outcomes here.  Forget about whether something is a conditional BP or not, or even how the implementation will work.  I want to know (a) exactly how the debugger text commands should look, and (b) exactly what the desired outcome will be from that command.  Once we have a clear idea of __what__ it is we want, then we can actually implement it that way.

Sorry to sound grumpy, but I get easily confused when stuff goes all over the place.  So let's stick to the _what_ first, and worry about the _how_ later.

**Comment by andrew-davie** [comment]
_"I want to know (a) exactly how the debugger text commands should look, and (b) exactly what the desired outcome will be from that command."_

1. break <label>

This behaves exactly as the break command worked in previous versions of Stella. But behind the scenes, it places multiple CBP (one per bank) instead of a single unconditional BP.

2. break <label> <bank>

This sets a breakpoint for a single bank/label. It is implemented as a CBP.

3. UI still shows red dot for when there's a BP, but it is always a CBP with a bank, so will only show up if the bank matches the CBP bank.

That's about it.  That's what I want.

**Comment by andrew-davie** [comment]
_Can you double check this, please? I cannot reproduce that bug._

Checked, Reproducible. I can send you the ROM if you wish to check that.
Again, load the ROM in stella, trigger the debugger.
Type 'break Reset' and it adds a breakpoint in bank B
Type 'break Reset F' and it adds a breakpoint in bank F (as expected)
Manually switch disassembly/view to bank F and confirm breakpoint is displayed at 'Reset' label.
type 'reset'
type 'run'
no breakpoint hit. 'Reset' label is actually pointed to by the reset vector of the ROM. This is absolutely positively guaranteed to be executed.

AH!  OK, after typing the above I thought - maybe it's the CLI setting that's an issue. So, to test, I un-clicked the BP in the UI, and then added it again with a click (no CLI). "reset" it triggers. So, setting BP via UI *is* working. Setting via CLI is not.

**Comment by andrew-davie** [comment]
I was wrong. Apologies.
When I type 'reset' I saw nothing happen, but actually the breakpoint WAS triggered.
But the disassembly view did not position on the breakpoint so I thought it hadn't.
So, I'll file that as a different bug.

**Comment by thrust26** [comment]
Was the breakpoint at a valid address? It will only display a breakpoint at the beginning of an instruction, not in the middle.

**Comment by thrust26** [comment]
> 1. break label
> 
> This behaves exactly as the break command worked in previous versions of Stella. But behind the scenes, it places multiple CBP (one per bank) instead of a single unconditional BP.

Can you give me an example why this should place multiple CBPs, please ? I can barley understand the reason for one "bank global" CBP, which can be enabled and disabled as a whole. Much less a breakpoint which is enabled "bank global" but can be disabled bank by band.

**Comment by andrew-davie** [comment]
_"Can you give me an example why this should place multiple CBPs, please ? I can barley understand the reason for one "bank global" CBP, which can be enabled and disabled as a whole. Much less a breakpoint which is enabled "bank global" but can be disabled bank by band."_

Most common usage will be that the person doing the debugging will NOT know the bank of something but WILL know the label.  So, I know that I want to break on "NewFrame", for example. So, I type "break NewFrame".  

IF the system adds N breakpoints for me, each a CBP one for each bank, I am now **guaranteed** that I will get that break happening. This is "old behaviour" without having old-style unconditional BPs.

Consider BoulderDash, with identical code in multiple banks. If I want to break on a label in that code (say, "ScreenDraw"), I don't want to have to hunt for every bank it's in. I just want to break on the label. Again, this is where CBP one in each bank is handy. If the system breaks on one I don't want/need, it's a simple UI click on the red circle to turn that one off. It's much easier to remove one that has happened that you DON'T want, than to hunt banks to add one that you DO.

And you are slightly misunderstanding. I am NOT proposing that you **disable** as a whole. Only enable. You enable via "break label" but to disable, you have to use the UI.  Having said that, it would be nice to have a command to remove a break, in which case "removebreak label" should remove all of 'em :)

And finally, there is "backwards compatibility".  I wanted anyone typing "break label" to have exactly the same behaviour as previous versions of stella. And that is, the debugger will halt when label's PC is reached *no matter what bank is active*.

**Comment by andrew-davie** [comment]
Embarrassing, but we’re all friends here… right?
The issue was that the debugger GUI defaulted to one line larger than my screen could display - and I was not aware of this.
So, yes the breakpoint was at a valid address - but the break happened and positioned the break line on that last line of the display. So I did not see it.

**Comment by thrust26** [comment]
@andrew-davie So `break label`would create n breakpoints for n banks and `break f000`would create only one? Is that what you want?

**Comment by andrew-davie** [comment]
> So break labelwould create n breakpoints for n banks and break f000would create only one? Is that what you want?

I honestly don't care on that one, as I rarely use the address form. But that would seem like a reasonable way to go if that works for you.

**Comment by andrew-davie** [comment]
Another option is to have 

break Label ALL

That is, where the bank number goes, you can type ALL to have a breakpoint in every bank at the label's PC address. This does break backward compatibility but I would be happy with this as a solution if you don't want 'break Label' to add N breaks.

**Comment by thrust26** [comment]
@andrew-davie 
I am currently working on this. Since the parser cannot use `ALL` (only numbers), we have to use `-1`. Weather this indicates toggling in all banks (what I would prefer) or limits to the current one (like you prefer), or there will be separate command is undecided yet. But there will be an option to set breaks in the current bank only, and another one to set in all banks.

But here is another thing:
The real 6507 addresses are only 13 bits (addr & 0x1fff). So we have at least the following options for e.g. `break f000`:
- set break only for f000
- set 8 breaks for 1000, 3000, ... f000
- set one break for 1000 (= f000 & 1fff, first mirror)
- any more???

Problems:
- In the first case, the debugger will only break if the PC has the correct 16 bit address. But the **full** PC address is only updated in case of `JSR`, `JMP`, `RTS` and `BRK`. Else the high bits may result from a previous bank with a different `ORG` and the break may never trigger.
- The seconds option will get the breakpoint, but it will clutter the command result and 'listbreaks'. And e.g. `break 5000` will clear all breaks created by `break f000`.
- The last option may be irritating (or maybe not), because the command will be reported and listed as `break 1000`. We could store the original address for reporting and listing, but internally it will always be be `1000`. So e.g. `break 5000` would **clear** it. 

The root problem are the upper three bits of the address. In reality they do not exist. But we developers got used to them and also used and still use them to workaround our tool's (e.g. DiStella) flaws. 

BTW: If we decide for one thing or the other we should make this consistent for the debugger.

**Comment by DirtyHairy** [comment]
I'd mask bits 15-13. If I understand correctly, this is option 3 😏

**Comment by thrust26** [comment]
Correct. I am leaning towards it too. But I am not exactly happy with it.

**Comment by thrust26** [comment]
I went for option 3 and also reimplemented the breakpoints which now use a hash map. So there are no slowdowns, even with a **LOT** of breakpoints.

As of now -1 for the bank enables a bp in all banks. This may change.

**Comment by andrew-davie** [comment]
Thank you, I will try and test soon.
"We" could always have a different named command to distinguish the two.

**Comment by thrust26** [comment]
Agreed. 

IMO `break`should be reserved for the normal breakpoints, because it is logical to autofill the current bank like it does with the current address too. And it stays in sync with setting a breakpoint via mouse.

So I need a good name for putting a breakpoint in all banks. Then you will also be able to type just `goodnamehere` and get a breakpoint in all banks at the current address.

**Comment by andrew-davie** [comment]
> IMO `break`should be reserved for the normal breakpoints, because it is logical to autofill the current bank like it does with the current address too. And it stays in sync with setting a breakpoint via mouse.

This is **NOT** my preferred course of action, as I truly believe if I say "break address" then I should get a breakpoint no matter what bank happens to be switched in when the CPU hits that address. As stated earlier, if I want to break in a specific bank and only that bank, then I should specify it.

Having said that, and hopefully making it totally clear I do not think this is a good solution... how about using "breakall" as the command to implement it the way you want?

**Comment by thrust26** [comment]
> This is NOT my preferred course of action, as I truly believe if I say "break address" then I should get a breakpoint no matter what bank happens to be switched in when the CPU hits that address. As stated earlier, if I want to break in a specific bank and only that bank, then I should specify it.

Following that argument, `break` would have to set a breakpoint at every address. :smile: I prefer things to be consistent. 

To me,`breakall` sounds more like the solution you prefer.

**Comment by andrew-davie** [comment]
> Following that argument, `break` would have to set a breakpoint at every address. 😄 I prefer things to be consistent.
> 
> To me,`breakall` sounds more like the solution you prefer.

I believe "break" should maintain its existing behaviour and break at every *bank* with the given address. Yes. And if you don't want that behaviour, then specify the bank you want. Or, disable the breaks as they trigger in the banks you don't want them, via UI click. 

So no, "breakall" is **not** the solution I prefer. It is an ugly hack/compromise to get around the issues you are having with parsing numbers/text.

I prefer "break" to do what it has always done - break at every bank with the given address. And if that's not what the user wants, with the extra 'bank' option, you get the choice to limit it to one bank.  THAT's what I prefer.

**Comment by thrust26** [comment]
To me we fixed a long standing bug with `break` command and toggling in the ROM listing (thanks for remembering us). For `break` both parameters are optional and automatically filled with the current values if not provided. Just like when clicking into the ROM listing. So that is fully consistent behavior which is more relevant to me than emulating old behavior.

I suppose this is one of those case where we both cannot agree, except that we agree on that. :smile:

**Comment by andrew-davie** [comment]
I'm not really wanting to argue this much more. But I will emphasise what I think are the important points regarding using 'break' in the debugger with a bankswitched game.

Of prime importance is that it is possible, even **probable** that the person doing the debugging has absolutely no idea of the actual bank in which code is living. All they have is a label (say, the entry to a subroutine). So, let's say they want to put a breakpoint at the entry to the subroutine 'DrawTheScreen'.

How do they do that in the new system via your method?  They have to use 'breakall' - in other words, put a breakpoint in every single bank so that the 'DrawTheScreen' PC address will be caught.

In other words, what I expect to be the DEFAULT behaviour for most breakpoints usage when debugging (because you almost certainly do not know the bank), is now to use 'breakall'.  You have therefore changed usage from 'break' (which used to do this - that is, effectively put breaks in all banks) to a new one 'breakall'. What has that actually achieved? Nothing but confusing existing users.

I think it's extremely unlikely and very rare that anyone will actually type the bank number when entering a breakpoint. After all, there is no bank number available in the symbol table, unless you specifically add a label for it to tie label/bank together somehow. 

So to me, what you are coding for is the least used case of all. In fact the most used case is the previous behaviour (break on all banks), and the whole discussion was to allow those extra bank breakpoints to be ***removed*** if required. And that lead to having the CBPs added, and thus easy removal of the breaks that were not wanted.

In summary, then, I think your use-case is not well considered. The previous behaviour was the most likely scenario (and not, IMHO, a bug). It only needed the change to CPB to allow individual ones to be removed as required.

Perhaps we just use the debugger totally differently. In any case, hopefully I'll fall silent now and you can implement it however you think is best. I appreciate all your hard work on this!

**Comment by andrew-davie** [comment]
Here is another use-case, which I have just encountered. Your game crashes. Stella halts, and you can see addresses on the stack. F809. You want to put a breakpoint on that location - but you don't, of course, know which bank it was from. The most common requirement here would be to want a break in EVERY bank so you could find out where it came from.

**Comment by thrust26** [comment]
First, I get your points even though I seem to develop and debug differently (knowing my banks). So they are relevant to me. I hope for the same vice versa for my arguments.

The problems we are facing are a direct result of multiple complexities:

1. the 6507 address space is 13 bit only
2. for extended space, we need bankswitching
3. neither DASM nor DiStella can provide labels with the correct bank
4. the PC may or may not have the upper 3 bits set as expected, so breakpoints may be missed

Previously the `break` command and click did not care for any of them. And now we are trying to fix that. Providing a bank is only part of the fix (2.), reducing the break addresses to 13 bit fixes 1. + 4. But that makes 3. even worse, because the upper 3 bits have a meaning during development because they are often used to differentiate banks. I will come back to this.

When designing a UI, the complexity should be mitigated by the design. Making the UI consistent is a major point here. Keeping it backward compatible is another one. 

Setting and clearing a breakpoint via `break` command or via clicking into the ROM list behaved identical in previous versions. This is fine. Therefore we would break backward compatibility and consistency and increase complexity if we would make the behavior different now. And the `break` command itself would become inconsistent too: One argument would be defaulted to the current state and the other one not. So that's why I decided to provide the current address and bank if an argument is not given. 

I then added the option to break in all banks to suit your needs (even though you are not happy with it). But by now I realized that this is not sufficient and your suggestion **too**. When you are searching for a label and its bank, the upper 3 bits usually (depending on coding style) become important. When you use the upper 3 bits to differentiate banks and there are no more than 8 banks, each label becomes totally unique. And even for more banks, they are still quite unique. So there is no need to set breakpoints in all banks (if you assume that the PC is set correctly, see 4.). Instead the break command should use all 16 bits in this case.

Therefore, depending on the situation, it makes sense to search for 13 or 16 bit addresses. I do not like to handle this via extra parameter as it adds extra complexity to the UI. Therefore we need a new command for the latter (e.g. `break16` or `breaklabel`). This new command will not even need a bank, it will (usually) find the correct bank from the upper 3 bits and therefore will default to any bank. And when this breakpoint has been found, I plan that the user can convert it into a normal `break address bank` one easily (disable and re-enable breakpoint). And last not least, this will also avoid cluttering the breakpoint list with unnecessary or even invalid (middle of instruction or data) breakpoints.

So that's my plan, the best solution I found by now. But I am open for more arguments and suggestions as long as my points (especially UI consistency) given above are considered too.

**Comment by thrust26** [comment]
Any more feedback on this one? Else I will close it in some days.

**Comment by andrew-davie** [comment]
Initial test has me confused. I have 3E bankswitching, and put a breakpoint on a label I *KNOW* is in bank $F. "Reset". That's the start vector entry point to the game. So, 'breaklabel Restart' and then 'run'. It breaks, but instead of in bank $F it's showing me a *correct disassembly* in bank $B.  I switch the bank to anything via the manual bank-select droplist in the 3E tab, and nothing changes.

I think the issue here is that the fixed bank ($F) is NOT one of the selectable banks - that is, the disassembly is showing the non-bankswitched code, but the title of it is as if it's a bank. Perhaps it would be useful to change the wording to include bank # ($F000-$F800) + fixed ($f800-$FFFF) or at least something. I know, this was my misunderstanding, but still...

Back to 'breaklabel'... it worked as advertised for a label in the fixed bank.
It did not work for me in a label in code that is copied to RAM. Specifically, the ORIGIN of the code was set to 0 (hence the label symbol is located at $000B). This code is copied into RAM which lives at $f800, so the PC address would be $f809, but the label is $0009, so there's that.
I would be happy to relocate the code/assembly which shouldn't make much difference, I think, but at least here is an example where using breaklabel does not work for the intended use-case, that is copying code to multiple banks and breaking on the label address.
I know this is a marginal case, and perhaps it could be argued that my code should not have an origin at 0. Let's discuss?

**Comment by thrust26** [comment]
I have to check the first case. Actually I am not quite clear what happens. Does it break at the wrong place or is the disassembly shown just the wrong one? Or is just the bank number display "wrong"?

"breaklabel" will of course only work correctly if the origin on assembly matches the one on execution. There will be cases were this problem circumvented and maybe there are other cases where this turns out difficult to impossible. The complexity strikes again here and probably we have to find a good compromise here. E.g. with 1K slice sizes, code with was assembled in slice 3, could go to slice 0, 1, 2 or 3 (even changing and/or multiple ones). There is no way for Stella to know this.

In your case, a correct ORG should fix the problem.

**Comment by andrew-davie** [comment]
The first case:

It broke in the correct place, say $FC2F.  That's in bank $F (fixed).
But the bank being displayed was (say) $B.  Even when I went to the 3E tab and manually switched banks, it **still** displayed the correct disassembled code, because no matter what bank is switched via $3F, the fixed bank is static.  So, code in $Fc2F is always visible.
It was just a misunderstanding on my part on what I was seeing. I thought that since it said "bank $b" I was actually looking at code in bank $B. But bank $B was switched into address range $f000-$F7FF, and of course the fixed bank is switched to $F800-$FFFF so of course I always see code at $FC2F, no matter what bank is switched in. In that case, I was suggesting that the title to the window could perhaps state that I am actually seeing the switched bank AND the fixed bank.

**Comment by thrust26** [comment]
Yes, that's irritating. For a general solution, we would have to either list all slices (only two in your case). IMO that's probably irritating as well. 

Or we detect the bank from the address and react accordingly. So when it breaks in the first slice(s), display "bank #N" from that slice, else we display e.g. "bank #15" or "fixed bank".

**Comment by andrew-davie** [comment]
How about put a little colour box next to the bank # display in the title, and use that colour as the background for the displayed bank ONLY. Thus, if you see a background of a different colour, you know it's NOT the bank displayed.  Also, when you click anywhere in the disassembly window, the bank # should update to show correctly what bank is currently being seen at that point/line?

**Comment by thrust26** [comment]
That would probably require a lot of work. The disassembler and the breakpoint handling are in very different places of the code. Making them interact would need quite some refactoring.

**Comment by andrew-davie** [comment]
> That would probably require a lot of work. The disassembler and the breakpoint handling are in very different places of the code. Making them interact would need quite some refactoring.

But the UI already has the ability to know what address a click is. You can set/clear breakpoints on any line, and the current PC line is highlighted... so clearly the address of each line is available. Given that, it would be pretty straightforward in the draw of the disassembly to change the background colour based on address on that line...?

**Comment by thrust26** [comment]
Yes, that would be possible. But that wouldn't be sufficient.

The root problem is, that there is no interface in the `Cartridge` class which provides information about the memory layout of the 4K. So we do not know the detailed bank slice sizes in the debugger. This is handled internally in the class and in the according widget classes. That would be ~100 classes affected.

**Comment by thrust26** [comment]
@andrew-davie 
I just enhanced the bank returned for some cart classes (incl. 3E). This now considers the address. This should help with the breakpoints (breaking and displaying better).

**Comment by andrew-davie** [comment]
> @andrew-davie
> I just enhanced the bank returned for some cart classes (incl. 3E). This now considers the address. This should help with the breakpoints (breaking and displaying better).

TY!

**Comment by thrust26** [comment]
I suppose this is as good as it gets for now.
